### ISSUE TITLE
File handles not released results in rclone crash

### ISSUE THREAD

**Issue by litemotiv** [rclone/rclone#3547]
System: Debian 64bit stable
Rclone: v1.49.0-071-gab89e939-beta
Remote: Dropbox
Command-line:  `rclone sync -vv -L --log-file=log.txt --tpslimit=5 <source> dropbox:<dest>`

Running sync results in certain file transfers erroring out with too_many_write_operations:

```
2019/09/17 14:06:43 DEBUG : multisite/wp-content/themes/digikit/vendor/mollie/mollie-api-php/examples/shipments/update-shipment.php: Unchanged skipping 
2019/09/17 14:06:43 DEBUG : pacer: low level retry 1/1 (error too_many_write_operations/..)
2019/09/17 14:06:43 DEBUG : pacer: Rate limited, increasing sleep to 20ms 
2019/09/17 14:06:43 DEBUG : multisite/wp-content/uploads/sites/66/2014/01/Dorpenlab-Hornhuizen-2.jpg: Received error: upload failed: too_many_write_operations/.. - low level retry 1/10 
2019/09/17 14:06:43 DEBUG : pacer: Reducing sleep to 15ms
```

File handles of transfers with errors are then not released, e.g:

```
# lsof -c rclone | wc -l
742
```

When the maximum number of open handles is reached rclone will crash with "Too many open files".

**Comment by ncw** [comment]
That looks like it is a bug!

I've managed to reproduce it.

This was introduced sometime between v1.48 and v1.49 in

commit 53a1a0e3efb2cf04aace48537a7691ada00e988b
Author: Aleksandar Jankovic <office@ajankovic.com>
Date:   Mon Jul 22 21:11:46 2019 +0200

    accounting: add reference to completed transfers
    
    Add core/transferred call that lists completed transfers and their
    status.

I haven't figured out exactly why yet but I will soon!

**Comment by ncw** [comment]
OK have a go with this - I think it should fix the problem

https://beta.rclone.org/branch/v1.49.0-075-gba6a7e30-fix-3547-fh-leak-beta/ (uploaded in 15-30 mins)

**Comment by litemotiv** [comment]
> OK have a go with this - I think it should fix the problem
> 
> https://beta.rclone.org/branch/v1.49.0-075-gba6a7e30-fix-3547-fh-leak-beta/ (uploaded in 15-30 mins)

Thanks Nick! The latest version definitely seems to improve things, although there might still be a small (related?) issue left. I'm still testing, but on the first run there were a large number of file transfers without adding open file handles, so that's good. 

At a certain point the sync seemed to stay stuck and stalled the queue:

```
2019/09/18 19:04:19 INFO  : multisite/wp-content/uploads/sites/113/2014/05/3-Park-Presikhaaf1-280x187.jpg: Copied (new)
2019/09/18 19:04:19 INFO  : multisite/wp-content/uploads/sites/113/2014/05/3-Park-Presikhaaf1-300x200.jpg: Copied (new)
2019/09/18 19:04:19 INFO  : multisite/wp-content/uploads/sites/113/2014/05/3-Park-Presikhaaf1-585x187.jpg: Copied (new)
2019/09/18 19:04:20 INFO  : multisite/wp-content/uploads/sites/113/2014/05/3-Park-Presikhaaf1-e1402236174182.jpg: Copied (new)
2019/09/18 19:04:20 INFO  : multisite/wp-content/uploads/sites/113/2014/05/3-Park-Presikhaaf1.jpg: Copied (new)
2019/09/18 19:04:20 INFO  : multisite/wp-content/uploads/sites/113/2014/05/3-Park-Sonsbeek-300x200.jpg: Copied (new)
2019/09/18 19:04:21 INFO  : multisite/wp-content/uploads/sites/113/2014/05/3-Park-Sonsbeek-280x187.jpg: Copied (new)
2019/09/18 19:04:21 INFO  : multisite/wp-content/uploads/sites/113/2014/05/3-Park-Sonsbeek-585x187.jpg: Copied (new)
2019/09/18 19:04:22 INFO  : multisite/wp-content/uploads/sites/113/2014/05/3-Park-Sonsbeek.jpg: Copied (new)
2019/09/18 19:04:32 INFO  :
Transferred:       49.583M / 714.847 MBytes, 7%, 230.098 kBytes/s, ETA 49m20s
Errors:                 0
Checks:            166726 / 166726, 100%
Transferred:          505 / 10517, 5%
Elapsed time:     3m40.6s
Transferring:
 * multisite/wp-content/u…-Sonsbeek1-280x187.jpg:200% /27.923k, 5.072k/s, -
 * multisite/wp-content/u…-Sonsbeek1-300x200.jpg:  0% /31.452k, 0/s, -
 * multisite/wp-content/u…-Sonsbeek1-585x187.jpg:  0% /52.271k, 0/s, -
 * multisite/wp-content/u…5/3-Park-Sonsbeek1.jpg:  0% /341.006k, 0/s, -

2019/09/18 19:05:32 INFO  :
Transferred:       49.583M / 714.847 MBytes, 7%, 180.906 kBytes/s, ETA 1h2m45s
Errors:                 0
Checks:            166726 / 166726, 100%
Transferred:          505 / 10517, 5%
Elapsed time:     4m40.6s
Transferring:
 * multisite/wp-content/u…-Sonsbeek1-280x187.jpg:200% /27.923k, 102/s, -
 * multisite/wp-content/u…-Sonsbeek1-300x200.jpg:  0% /31.452k, 0/s, -
 * multisite/wp-content/u…-Sonsbeek1-585x187.jpg:  0% /52.271k, 0/s, -
 * multisite/wp-content/u…5/3-Park-Sonsbeek1.jpg:  0% /341.006k, 0/s, -

2019/09/18 19:06:32 INFO  :
Transferred:       49.583M / 714.847 MBytes, 7%, 149.044 kBytes/s, ETA 1h16m10s
Errors:                 0
Checks:            166726 / 166726, 100%
Transferred:          505 / 10517, 5%
Elapsed time:     5m40.6s
Transferring:
 * multisite/wp-content/u…-Sonsbeek1-280x187.jpg:200% /27.923k, 2/s, -
 * multisite/wp-content/u…-Sonsbeek1-300x200.jpg:  0% /31.452k, 0/s, -
 * multisite/wp-content/u…-Sonsbeek1-585x187.jpg:  0% /52.271k, 0/s, -
 * multisite/wp-content/u…5/3-Park-Sonsbeek1.jpg:  0% /341.006k, 0/s, -
```
These files also stayed visible as open handles with 'wc -l'. The first stuck file seems to show 200% transferred which may or may not be related. 

I could reproduce this with a second run in the terminal, but for some reason i haven't been able to reproduce it in two runs with logging enabled. If you have have any clues about this or need more information please let me know.

**Comment by ncw** [comment]
If you can get it to get stuck again, can you 

    kill -QUIT pid

Rclone and post the traceback - that should need enough to work out what is going on.

**Comment by litemotiv** [comment]
> If you can get it to get stuck again, can you
> 
> ```
> kill -QUIT pid
> ```
> 
> Rclone and post the traceback - that should need enough to work out what is going on.

Currently it's been running for over an hour without problems and without any open files, so that looks great. I think the original bug can be marked as fixed, shall i open a new one in case it gets stuck again and i can provide you with a backtrace, or do you prefer to keep this one open?

Thanks Nick!

**Comment by ncw** [comment]
I can't think of any way that this patch would cause a lockup...  Though it might do - concurrent programming is hard!

If you see another lockup then please open a new issue.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.50
